### DISCUSSION TITLE
Bypassing automated crawler detection by Firewalls

### DISCUSSION THREAD

**Discussion by dnmahendra** [unclecode/crawl4ai#507] [Forums - Q&A]
Is there a solution for websites behind WAFs like PerimeterX, Cloudflare, Akmai etc.?

**Comment by unclecode** [comment]
Thank you for raising this important question about bypassing Web Application Firewalls (WAFs) like PerimeterX, Cloudflare, and Akamai. While completely bypassing advanced WAFs can be challenging and may raise ethical concerns, Crawl4ai already has several features that can help mitigate some basic anti-bot measures:

1. **User-Agent Customization**: You can set a custom User-Agent to mimic legitimate browser requests:
   ```python
   crawler.crawler_strategy.update_user_agent("Your Custom User-Agent")
   ```

2. **Proxy Support**: Use proxies to distribute requests across different IP addresses:
   ```python
   crawler = AsyncWebCrawler(proxy="http://your-proxy-url:port")
   ```

3. **JavaScript Execution**: Crawl4ai can execute JavaScript, which is crucial for rendering dynamic content:
   ```python
   result = await crawler.arun(url="https://example.com", js_code="Your JavaScript Code")
   ```

4. **Session-Based Crawling**: Maintain sessions to mimic human-like browsing behavior:
   ```python
   result = await crawler.arun(url="https://example.com", session_id="unique_session_id")
   ```

5. **Custom Headers**: Set custom headers to include necessary cookies or authentication information:
   ```python
   crawler.crawler_strategy.set_custom_headers({"Cookie": "your_cookie_value"})
   ```

These features can help in many cases, but not specifically for WAFs, but for other applications we're considering features. like enhanced browser fingerprinting, CAPTCHA, human-like behaviour and more.

I'd love to hear more about your specific use case. Are there particular websites or WAFs you're encountering issues with?

**Comment by DungNguyen83** [comment]
could be nice if you can confirm your code can scrapy this page https://www.realestate.com.au/property-house-qld-indooroopilly-145860356? @unclecode
